### ISSUE TITLE
Errors due to HPackDecoder visibility when using grpc 1.43.0

### ISSUE THREAD

**Issue by mad-ned** [grpc/grpc-java#9253]
Hi, I am trying to upgrade our grpc java code from 1.25 up to 1.43, but I ran into the  (previously discussed) issue about HPackDecoder being moved to private it seems.  I am using the current set of grpc packages version 1.43.0 with my maven build including grpc-netty-shaded, and I get stack trace below. if I decompile the HpackDecoder it is indeed private, and the DefaultHttp2HeadersDecoder code seems to be still trying to create it.  Any ideas of what I have to do /change to get it to work?  

```
at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2Exception.newStatic(Http2Exception.java:72)
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.HpackDecoder.<clinit>(HpackDecoder.java:55)
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2HeadersDecoder.<init>(DefaultHttp2HeadersDecoder.java:73)
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2HeadersDecoder.<init>(DefaultHttp2HeadersDecoder.java:59)
        at io.grpc.netty.shaded.io.grpc.netty.GrpcHttp2HeadersUtils$GrpcHttp2ServerHeadersDecoder.<init>(GrpcHttp2HeadersUtils.java:60)
        at io.grpc.netty.shaded.io.grpc.netty.NettyServerHandler.newHandler(NettyServerHandler.java:166)
```

**Comment by ejona86** [comment]
Please include the actual error message you saw. You have the stack trace, but removed the exception itself.

Where was this previously discussed?

**Comment by mad-ned** [comment]
Thanks for the reply @ejona86 .  There was apparently a back-and-forth about making HPackdecoder private in 2017, as discussed here:  https://github.com/netty/netty/issues/6579  My reading of this and the other threads mentioned here is that netty made HPackdecoder private, but then grpc was supposed to do something to adjust for that.  

Full stack is below.   we are using Java 8 and Maven build, and generally following grpc.io advice on the pom file dependencies - including adding grpc-netty-shaded, grpc-protobuf, grpc-api etc, all version 1.43.  We construct a grpc server in our Java code using ServerBuilder, leading to this exception shortly thereafter. 

Thanks again for any help -Ned 

java.lang.NoSuchMethodError: java.lang.Exception: method <init>(Ljava/lang/String;Ljava/lang/Throwable;ZZ)V not found
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2Exception.<init>(Http2Exception.java:82)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2Exception.<init>(Http2Exception.java:34)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2Exception$StacklessHttp2Exception.<init>(Http2Exception.java:323)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2Exception.newStatic(Http2Exception.java:72)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.HpackDecoder.<clinit>(HpackDecoder.java:55)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2HeadersDecoder.<init>(DefaultHttp2HeadersDecoder.java:73)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2HeadersDecoder.<init>(DefaultHttp2HeadersDecoder.java:59)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.grpc.netty.GrpcHttp2HeadersUtils$GrpcHttp2ServerHeadersDecoder.<init>(GrpcHttp2HeadersUtils.java:60)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.grpc.netty.NettyServerHandler.newHandler(NettyServerHandler.java:166)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.grpc.netty.NettyServerTransport.createHandler(NettyServerTransport.java:263)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.grpc.netty.NettyServerTransport.start(NettyServerTransport.java:131)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.grpc.netty.NettyServer$1.initChannel(NettyServer.java:290)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:129)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:112)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.callHandlerAdded(AbstractChannelHandlerContext.java:938)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:609)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.access$100(DefaultChannelPipeline.java:46)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline$PendingHandlerAddedTask.execute(DefaultChannelPipeline.java:1463)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.callHandlerAddedForAllHandlers(DefaultChannelPipeline.java:1115)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.invokeHandlerAddedIfNeeded(DefaultChannelPipeline.java:650)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:514)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:429)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:486)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:387)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at io.grpc.netty.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)[106:com.<somewhere>:15.22.0.SNAPSHOT]
        at java.lang.Thread.run(Thread.java:748)[:1.8.0_202-CDS_2021_b1]

**Comment by ejona86** [comment]
HpackDecoder looks fine there. The problem is with Http2Exception. Note that there isn't an error causing the creation of Http2Exception; they are `static` exceptions that are reused.

The stack trace is a failure during construction:
https://github.com/netty/netty/blob/b78d8f2abda50fbc0ff3f177d72d1a0b05b5ff41/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java#L82

It appears to be complaining that `new Exception(String, Throwable)` is unavailable. That is... strange. But that's what it says. That constructor was added in Java 4. Note that Http2Exception.java:63 should be the line things are actually failing on, but it isn't shown for some reason (maybe some Java implementation detail about super()s in constructors).

Something seems busted with your JDK, but I have no clue what could cause this.

**Comment by ejona86** [comment]
Not any more we can do here, so closing. The method mentioned in NoSuchMethodError clearly exists in our build. gRPC looks to be getting repackaged into another jar, so it seems something is broken with the build of that jar. But I can't tell you how it is misconfigured.

**Comment by ejona86** [comment]
If you find some additional clues, comment, and we can reopen.

**Comment by mad-ned** [comment]
Thanks @ejona86 .  I meant to comment earlier.  You are right about it being related to the Exception class not being found, I had ignored the last few frames in the exception handler believing this was just the way an earlier exception issue about the HPackDecoder was being reported.  But it does appear that Exception itself is not being found.  I think I may have figured out why as well - we use Apache karaf runtime and for whatever reason, Apache overrides java.lang.Exception with its own class.  For some reason our Maven plugin with grpc does not pick that up, it is a transitive dependency from another package.  I have added it as an explicit one but still no luck.  It seems odd that the original java.lang.Exception is also not found in this case.  I have tried hacking the plugin install dir to move aside the apache plugin for this and the exception error does go away. (but this causes later problems downstream).  

Long way of saying, I don't think its your issue.  But it may be a general usability issue about using karaf with grpc, I don't know.
Thanks
Ned
